tokio-util/src/sync/cancellation_token.rs

```diff
@@ -55,30 +50,24 @@ use guard::DropGuard;
 /// }
 /// ```
 pub struct CancellationToken {
-    inner: NonNull<CancellationTokenState>,
+    inner: Arc<implementation::TreeNode>,
 }
 
-// Safety: The CancellationToken is thread-safe and can be moved between threads,
-// since all methods are internally synchronized.
-unsafe impl Send for CancellationToken {}
-unsafe impl Sync for CancellationToken {}
-
 /// A Future that is resolved once the corresponding [`CancellationToken`]
 /// was cancelled
 #[must_use = "futures do nothing unless polled"]
 pub struct WaitForCancellationFuture<'a> {
     /// The CancellationToken that is associated with this WaitForCancellationFuture
-    cancellation_token: Option<&'a CancellationToken>,
-    /// Node for waiting at the cancellation_token
-    wait_node: ListNode<WaitQueueEntry>,
-    /// Whether this future was registered at the token yet as a waiter
-    is_registered: bool,
+    _cancellation_token: CancellationToken,
+    /// Future, to wait for cancellation
+    future: Option<Pin<Box<tokio::sync::futures::Notified<'a>>>>,
```

Comment:
We do not need to heap allocate here.
```suggestion
    future: Option<tokio::sync::futures::Notified<'a>>,
```

Reply:
I wasn't able to get it to compile without Pin<Box<>>. I'll try again later.

Reply:
I believe to solve this we would have to implement [projections](https://doc.rust-lang.org/std/pin/#projections-and-structural-pinning).

Reply:
I read more about projections and got to the conclusion that it's difficult. I realized we already use a projection crate, `pin-project-lite`. Sadly, this one doesn't work (to my understanding) with the `WaitForCancellationFuture`, because we wrap our nested future in an `Option`. Which brings us again to the same problem that we have a `Pin<&mut Option<Future>>`, but we need a `Pin<&mut Future>` to actually do the polling.
Will read about it some more, but I'm still wondering how problematic a `Pin<Box<>>` would really be.

Reply:
Never mind. `.as_pin_mut()` does exactly that.

Updated.